plugins/inputs/spark/spark.go

```diff
@@ -0,0 +1,441 @@
+}
+
+func newSparkMetric(host string, metric string,
+	acc telegraf.Accumulator) *sparkMetric {
+	return &sparkMetric{host: host, metric: metric, acc: acc}
+}
+
+func newYarnMetric(host string, acc telegraf.Accumulator) *yarnMetric {
+	return &yarnMetric{host: host, acc: acc}
+}
+
+func addValuesAsFields(values map[string]interface{}, fields map[string]interface{},
+	mname string) {
+	for k, v := range values {
+		if v != nil {
+			fields[mname+"_"+k] = v
+		}
+	}
+}
+
+func parseJmxMetricRequest(mbean string) map[string]string {
+	tokens := make(map[string]string)
+	classAndPairs := strings.Split(mbean, ":")
+	if classAndPairs[0] == "metrics" {
+		tokens["class"] = "spark_jolokiaMetrics"
+	} else if classAndPairs[0] == "java.lang" {
+		tokens["class"] = "java"
+	} else {
+		return tokens
+	}
+
+	pair := strings.Split(classAndPairs[1], "=")
+	tokens[pair[0]] = pair[1]
+
+	return tokens
+}
+
+func addTokensToTags(tokens map[string]string, tags map[string]string) {
+	for k, v := range tokens {
+		if k == "name" {
+			tags["mname"] = v // name seems to a reserved word in influxdb
+		} else if k == "class" || k == "type" {
+			continue // class and type are used in the metric name
+		} else {
+			tags[k] = v
+		}
+	}
+}
+
+func addJavaMetric(class string, c javaMetric,
+	values map[string]interface{}) {
+
+	tags := make(map[string]string)
+	fields := make(map[string]interface{})
+	tags["spark_host"] = c.host
+	tags["spark_class"] = class
+
+	if class == "spark_Threading" {
+		list := []string{"PeakThreadCount", "CurrentThreadCpuTime", "DaemonThreadCount", "TotalStartedThreadCount", "CurrentThreadUserTime", "ThreadCount"}
+		for _, value := range list {
+			if values[value] != nil {
+				fields[value] = values[value]
+			}
+		}
+	} else {
+		for k, v := range values {
+			if v != nil {
+				fields[k] = v
+			}
+		}
+	}
+	c.acc.AddFields(class, fields, tags)
+
+}
+
+func (j javaMetric) addTagsFields(out map[string]interface{}) {
+	fmt.Println(out["request"])
+	request := out["request"].(map[string]interface{})
+	var mbean = request["mbean"].(string)
+	var mbeansplit = strings.Split(mbean, "=")
+	var class = mbeansplit[1]
+
+	if valuesMap, ok := out["value"]; ok {
+		if class == "Memory" {
+			addJavaMetric("spark_HeapMemoryUsage", j, valuesMap.(map[string]interface{}))
+		} else if class == "Threading" {
+			addJavaMetric("spark_Threading", j, valuesMap.(map[string]interface{}))
+		} else {
+			fmt.Printf("Missing key in '%s' output response\n%v\n",
+				j.metric, out)
+			return
+		}
+
+	}
+}
+
+func addSparkMetric(mbean string, c sparkMetric,
+	values map[string]interface{}) {
+
+	tags := make(map[string]string)
+	fields := make(map[string]interface{})
+
+	tokens := parseJmxMetricRequest(mbean)
+	addTokensToTags(tokens, tags)
+	tags["spark_host"] = c.host
+
+	addValuesAsFields(values, fields, tags["mname"])
+	c.acc.AddFields(tokens["class"]+tokens["type"], fields, tags)
+
+}
+
+func (c sparkMetric) addTagsFields(out map[string]interface{}) {
```

Comment:
It doesn't matter. Still for best practices I have converted it to pointer receiver